versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestFetchValuesUsingOpenAddressing.java

```diff
@@ -199,6 +165,105 @@ public void segmentPrefetching() {
+    "2, 1, 4",
+    "3, 2, 4",
+    // Wraparound from the second segment to and earlier index in the same segment.
+    // Note: the second segment is scanned twice
+    "5, 4, 4",
+    "7, 5, 4",
+    // Wraparound from the last segment to the second.
+    "15, 6, 2",
+    // Wraparound from the last segment to the first.
+    "15, 0, 1",
+    "14, 3, 1",
+  })
+  public void keyCollision(int naturalPos, int movedPos, int movedKeyFoundInRound) {
+    Function<Integer, Key> bucketToKey =
+        bucket -> {
+          // Simulate key collision by artificially crafting their hash codes instead of bothering
+          // with real insertion.
+          String name = "ordinary-" + bucket; // no hash collision
+          int hash = bucket;
+          if (bucket == movedPos) {
+            name = "moved-" + bucket;
+            // hash collides with `naturalPos` and this key is stored at `movedPos`
+            hash = naturalPos;
+          }
+          if (bucket == naturalPos) {
+            // hash collision, the key keeps its natural spot (the name helps debugging)
+            name = "natural-" + bucket;
+          }
+          return new KeyWithSpecificHash(hash, ImmutableList.of(name));
+        };
+
+    // Note: test parameters depend on these two values
+    int segmentCount = 4;
+    int entrySize = 4;
+
+    CommitLogEntry entry = getCommitFixture(entrySize, segmentCount, bucketToKey);
+
+    // Validate lookup for all keys
+    for (int b = 0; b < segmentCount * entrySize; b++) {
+      Key key = bucketToKey.apply(b);
+      // Reset FetchValuesUsingOpenAddressing for each key for ease of validation
+      FetchValuesUsingOpenAddressing fetch = new FetchValuesUsingOpenAddressing(entry);
+      AtomicReference<KeyListEntry> hit = new AtomicReference<>();
+      // Simulate loading key lists in multiple rounds
+      for (int r = 0; r < segmentCount + 1; r++) {
+        // Note: Do not preload segments for each of validation
+        List<Hash> hashes = fetch.entityIdsToFetch(r, 0, ImmutableList.of(key));
+        hashes.forEach(
+            h -> {
+              int segment = Integer.parseInt(h.asString());
+              KeyListEntity keyListEntity =
+                  getKeyListEntity(segment * entrySize, segment, entrySize, bucketToKey);
+              fetch.entityLoaded(keyListEntity);
+            });
+
+        fetch.checkForKeys(r, ImmutableList.of(key), hit::set);
+        if (hit.get() != null) {
+          if (b == movedPos) { // The moved key requires loading extra segments
+            assertThat(r)
+                .describedAs("Expected round for key " + key)
+                .isEqualTo(movedKeyFoundInRound);
+          } else { // Ordinary keys are found on the first attempt
+            assertThat(r).describedAs("Expected round for key " + key).isEqualTo(0);
+          }
+          break;
+        }
+      }
+
+      // Make sure each key is found
+      assertThat(hit.get()).describedAs("Key " + key).isNotNull();
```

Comment:
Nit: can be continued w/ `.extracting(KeyListEntry::getKey).isEqualTo(key);` (+ removing the next line).
Maybe better: Use the `@ExtendWith(SoftAssertionsExtension.class)` + an instance field `@InjectSoftAssertions SoftAssertions softly` and use `softly.assertThat`. Helps a lot when using many assertThat in a single test

Reply:
Converted to `.extracting`